sdk/go/common/workspace/project.go

```diff
@@ -102,89 +102,118 @@ type PluginOptions struct {
 	Path    string `json:"path" yaml:"path"`
 }
 
-// PackageSpec defines the structured format for a package dependency
+// PackageSpec defines the structured format for a package dependency, as represented on disk.
+//
+// When marshaling, PackageSpec will be represented as <source>@<version> if that is
+// sufficient to represent the struct. Otherwise the full object will be used.
+//
+// PackageSpecs are unresolved, and must be resolved before usage.
 type PackageSpec struct {
-	Source     string   `json:"source" yaml:"source"`
-	Version    string   `json:"version,omitempty" yaml:"version,omitempty"`
-	Parameters []string `json:"parameters,omitempty" yaml:"parameters,omitempty"`
-}
-
-// packageValue can be either a string or a PackageSpec
-// This is a private implementation type that handles the dual-format parsing
-type packageValue struct {
-	// The underlying value, either a string or a PackageSpec
-	value any
-}
-
-// Spec returns the package as a PackageSpec if it's a PackageSpec,
-// or creates a simple PackageSpec from its string representation.
-func (pv *packageValue) Spec() PackageSpec {
-	switch v := pv.value.(type) {
-	case PackageSpec:
-		return v
-	case string:
-		// Check if the string contains a version specifier in the format "source@version"
-		parts := strings.Split(v, "@")
-		if len(parts) == 2 {
-			return PackageSpec{
-				Source:  parts[0],
-				Version: parts[1],
-			}
+	// The "name" of the plugin.
+	//
+	// Source may be one of:
+	// - A simple name, like "pkg"
+	// - A registry double or triple: "org/pkg", "source/org/pkg"
+	// - A git URL, "git://github.com/pulumi/pulumi-example/path"
+	// - A local path, like /usr/bin/pkg
+	Source string
+	// The version of the provider, may be Semver 2.0 or a git hash.
+	Version string
+	// Any parameters needed to configure the package.
+	Parameters []string
+
+	// if set will be used to validate the plugin downloaded matches. This is keyed by
+	// "$os-$arch", e.g. "linux-x64".
+	Checksums map[string][]byte
+
+	// The server to download the plugin from, if needed.
+	PluginDownloadURL string
+
+	// When marshaling, prefer to unmarshal without the <name>@<version> shorthand.
+	unmarshalledFromFull bool
+}
+
+type packageSpecMarshaller struct {
```

Comment:
```suggestion
type packageSpecMarshalled struct {
```

marshaller sounds like this is the type _doing_ the marshalling, but this is just the marshalled form of PackageSpec